examples/robot-assistant/wasm/src/element.rs

```diff
@@ -0,0 +1,344 @@
+use wasm_bindgen::prelude::*;
+use wasm_bindgen::JsCast;
+use web_sys::EventTarget;
+
+/// Wrapper for `web_sys::Element` to simplify calling different interfaces
+pub struct Element {
+    el: Option<web_sys::Element>,
+}
+
+impl From<web_sys::Element> for Element {
+    fn from(el: web_sys::Element) -> Element {
+        Element { el: Some(el) }
+    }
+}
+
+impl From<web_sys::EventTarget> for Element {
+    fn from(el: web_sys::EventTarget) -> Element {
+        let el = wasm_bindgen::JsCast::dyn_into::<web_sys::Element>(el);
+        Element { el: el.ok() }
+    }
+}
+
+impl From<Element> for Option<web_sys::Node> {
+    fn from(obj: Element) -> Option<web_sys::Node> {
+        if let Some(el) = obj.el {
+            Some(el.into())
+        } else {
+            None
+        }
+    }
+}
+
+impl From<Element> for Option<EventTarget> {
+    fn from(obj: Element) -> Option<EventTarget> {
+        if let Some(el) = obj.el {
+            Some(el.into())
+        } else {
+            None
+        }
+    }
+}
+
+impl Element {
+    // Create an element from a tag name
+    pub fn create_element(tag: &str) -> Option<Element> {
+        if let Ok(el) = web_sys::window()?.document()?.create_element(tag) {
+            Some(el.into())
+        } else {
+            None
+        }
+    }
+
+    pub fn qs(selector: &str) -> Option<Element> {
+        let body: web_sys::Element = web_sys::window()?.document()?.body()?.into();
+        let el = body.query_selector(selector).ok()?;
+        Some(Element { el })
+    }
+
+    /// Add event listener to this node
+    pub fn add_event_listener<T>(&mut self, event_name: &str, handler: T)
+    where
+        T: 'static + FnMut(web_sys::Event),
+    {
+        let cb = Closure::wrap(Box::new(handler) as Box<dyn FnMut(_)>);
+        if let Some(el) = self.el.take() {
+            let el_et: EventTarget = el.into();
+            el_et
+                .add_event_listener_with_callback(event_name, cb.as_ref().unchecked_ref())
+                .unwrap();
+            cb.forget();
+            if let Ok(el) = el_et.dyn_into::<web_sys::Element>() {
+                self.el = Some(el);
+            }
+        }
+    }
+
+    /// Delegate an event to a selector
+    pub fn delegate<T>(
+        &mut self,
+        selector: &'static str,
+        event: &str,
+        mut handler: T,
+        use_capture: bool,
+    ) where
+        T: 'static + FnMut(web_sys::Event),
+    {
+        let el = match self.el.take() {
+            Some(e) => e,
+            None => return,
+        };
+        if let Some(dyn_el) = &el.dyn_ref::<EventTarget>() {
+            if let Some(window) = web_sys::window() {
+                if let Some(document) = window.document() {
+                    // TODO document selector to the target element
+                    let tg_el = document;
```

Comment:
If I'm not mistaken, all of these `if let`'s will pretty much skip the whole function, is that right? If that's the case, you can avoid a significant amount of rightward-drift by changing the return type to `Option<()>` and simply using the `?` operator. So this would look more like this:

```rust

    pub fn delegate<T>(
        &mut self,
        selector: &'static str,
        event: &str,
        mut handler: T,
        use_capture: bool,
    ) -> Option<()> where
        T: 'static + FnMut(web_sys::Event),
    {
        let el = self.el.take()?;
        let dyn_el = &el.dyn_ref::<EventTarget>()?;
        let window = web_sys::window()?;
        let document = window.document()?;
        // ...
        Some(())
    }
```

Reply:
This piece of code will be replace by seed. I copy pasted this from wasm-bindgen source code and didn't do much investigation on it.